src/raft.rs

```diff
@@ -2457,4 +2559,53 @@ impl<T: Storage> Raft<T> {
         to_send.set_entries(req.take_entries());
         Some(to_send)
     }
+
+    /// Reduce size of 'ents' from uncommitted size.
+    pub fn reduce_uncommitted_size(&mut self, ents: &[Entry]) {
+        // fast path for non-leader endpoint
+        if self.state != StateRole::Leader {
+            return;
+        }
+
+        if !self.uncommitted_state.maybe_reduce_uncommitted_size(ents) {
+            // this will make self.uncommitted size not accurate.
+            // but in most situation, this behaviour will not cause big problem
+            warn!(
+                self.r.logger,
+                "try to reduce uncommitted size less than 0, first index of pending ents is {}",
+                ents[0].get_index()
+            );
+        }
+    }
+
+    /// Increase size of 'ents' to uncommitted size. Return true when size limit
+    /// is satisfied. Otherwise return false and uncommitted size remains unchanged.
+    /// For raft with no limit(or non-leader raft), it always return true.
+    pub fn maybe_increase_uncommitted_size(&mut self, ents: &[Entry]) -> bool {
+        // fast path for non-leader endpoint
+        if self.state != StateRole::Leader {
+            // fast path should not block 'append' operation on non-leader raft
+            return true;
+        }
+
+        self.uncommitted_state.maybe_increase_uncommitted_size(ents)
+    }
+
+    /// Compute current uncommitted log size. Only called by leader.
+    pub fn compute_uncommitted_size(&self) -> usize {
+        if self.uncommitted_state.is_no_limit() {
+            return 0;
+        }
+
+        // since uncommitted entries of leader won't be compacted,
+        // any error return by slice should cause panic
+        match self.raft_log.entries(self.raft_log.committed + 1, None) {
```

Comment:
I don't think it's a good idea. We have optimized hard for avoiding reading logs from hard disk when becoming leader. Perhaps we can keep the size limit a soft limit instead of a hard one. /cc @NingLin-P

Reply:
@NingLin-P @BusyJay just like `etcd`, we can simply set `uncommitted_size` to zero rather than re-computing it.

Reply:
We have `last_committed_index` tracing where `uncommitted_size` start counting, as long as they are consistent I think it is always okay, for example:
- `last_committed_index = last_index`, `uncommitted_size = 0`
- `last_committed_index = index of the first unstable log `, `uncommitted_size = unstable log szie`

Reply:
LGTM

Reply:
> We have `last_committed_index` tracing where `uncommitted_size` start counting, as long as they are consistent I think it is always okay, for example:
> 
> * `last_committed_index = last_index`, `uncommitted_size = 0`
> * `last_committed_index = index of the first unstable log `, `uncommitted_size = unstable log szie`

We still need to pick a solution @Dantelee